enh: scrolling view more intuitive guide

shun2wang opened the pull request:
Fix: https://github.com/jasp-stats/jasp-issues/issues/3300


I'm always annoyed by the default maximized windows, especially after the system memorizes my desktop window layout.I think it's also not good experience for users. 
so let's make scroll view more intuitive.

This is not the only solution, see also: https://ux.stackexchange.com/questions/38771/how-to-hint-that-theres-more-beyond-what-you-see-in-a-scroll-view

Looks now analysis form:
![image](https://github.com/user-attachments/assets/ca615992-4fd8-449b-a0a7-578d98b098ce)
menu panel:
![image](https://github.com/user-attachments/assets/cfc5fbbd-5635-4ba1-81f4-a163f555d63f)

shun2wang committed:
AnalysisForms scrolling guide better

shun2wang committed:
also add scrolling guide to custom menu

shun2wang committed:
disable maximized main window

shun2wang commented:
Or should it be a reusable component?

boutinb reviewed (commented):
I find it quite a good solution.
I would use the jaspTheme.fileMenuColorBackground for the CustomMenu.qml
and jaspTheme.analysisBackgroundColor for the AnalysisForms.qml

shun2wang commented:
>  I would use the jaspTheme.fileMenuColorBackground for the CustomMenu.qml and jaspTheme.analysisBackgroundColor for the AnalysisForms.qml

I tried it, but it seems that the color depth contrast at the bottom is not strong enough, and there is still the illusion that "there is no content below"?or maybe just like a shadow?

boutinb commented:
> > I would use the jaspTheme.fileMenuColorBackground for the CustomMenu.qml and jaspTheme.analysisBackgroundColor for the AnalysisForms.qml
> 
> I tried it, but it seems that the color depth contrast at the bottom is not strong enough, and there is still the illusion that "there is no content below"?or maybe just like a shadow?

I don't have strong preferences here....
I'm also a bit annoyed by the default maximized window, so I really prefer this way. @JorisGoosen , @EJWagenmakers ?

JorisGoosen commented:
Ill add a preference option for open maximized that is off by default

(Because I love the maximized opening, saves me maximizing it every time I use it)
(Should just show helpwindow and engine window inside mainwindow and its all perfect.)

JorisGoosen committed:
add qml widget ScrollMoreIndicator and change color to "shadow"

also make it become smaller smoothly as you scroll down, it just disappearing is ugly.

JorisGoosen committed:
also the top of the analyses and custommenu

JorisGoosen committed:
smoother

JorisGoosen committed:
make sure the preferences and the filemenu also get a scrollindicator

JorisGoosen committed:
also add indicators to modules/dataworkspace menu on the right

JorisGoosen committed:
also make the fileselectors have a nice indicator

JorisGoosen committed:
also add a setting for opening jasp maximized or not

Yes it maximizes JASP when enabling this option but given that this only happens for people who want JASP to start maximized I think this will be fine.

JorisGoosen commented:
Alright, this is almost ready for review again, by @boutinb  and by @shun2wang 

I did change the color to shadow, I think it looks pretty good in both light and dark theme.
While the previous color looked fine in darktheme it looked invisble in lighttheme.

I also think now we have scroll indicators everywhere?
The ribbon had them for a while already and I dont know what Im missing?

Aaah, the labeleditor!
Ok, ill add that.
Ah and the dropdowns...

Ok so maybe this is not ready for review yet.

JorisGoosen committed:
add scroll indicator to labelwindow and dropdown

also fixs some outlining issues in the labeleditor

JorisGoosen committed:
put lines around filter icon and column add button

also make column wider so german remove actually fits

JorisGoosen commented:
Ok, so I got a bit carried away, sorry for hijacking your PR @shun2wang 

But now we should have scroll indicators everywhere except the data.
Also it should make the labeleditor look a bit better (no horizontal scrollbar for instance).

shun2wang commented:
Behavior on drop-down now is it intentional? seems shadow always showing what ever if scrollbar is appeared.
![image](https://github.com/user-attachments/assets/14edbf26-5f7f-4018-bde7-b03ac8993d59)

shun2wang commented on Desktop/components/JASP/Widgets/FileMenu/DataLibrary.qml:
```diff
@@ -70,6 +70,33 @@ Item
+			top:			secondseparator.bottom
+			topMargin:		-secondseparator.height / 2
+			left:			parent.left
+			right:			parent.right
```
Should also set leftMargin? It seems like it should be anchored to the left side of the scrollbar, as it is done in the button? But it's not a big deal. I'm not really a stickler for this idea;-)

![image](https://github.com/user-attachments/assets/777cc0ed-7f78-4afd-b004-07ef7ced8d97)

JorisGoosen replied:
Yeah, it wouldve been prettier, but there are some other issues there as well.
Like, it wont scroll all the way down for some reason.

But I was a bit distracted by the labeleditor :p

Ill have a look later

JorisGoosen commented:
> Behavior on drop-down now is it intentional? seems shadow always showing what ever if scrollbar is appeared. ![image](https://private-user-images.githubusercontent.com/10348402/434879034-14edbf26-5f7f-4018-bde7-b03ac8993d59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5NjU3MTMsIm5iZiI6MTc0NDk2NTQxMywicGF0aCI6Ii8xMDM0ODQwMi80MzQ4NzkwMzQtMTRlZGJmMjYtNWY3Zi00MDE4LWJkZTctYjAzYWM4OTkzZDU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE4VDA4MzY1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhYmU0YzdiMjU5ZGVlYjgzZTllOWQ0YWEwYWJjYjA1NWMwYzM3YmQ3NjZmNzBlOTFhOTBlNjdhY2ExNGQ1NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5Cpv6HdMxaGTxXEE852WbohpK_9ZrNZC72EnVr-NYzU)

Ah looks like a bug!

shun2wang commented:
1. The gradient you defined seems to be from the middle to the sides😜. I think it would be better to have a gradient from one side to the other, and the color gets darker as it gets closer to the bottom (top).
2. Should we also ensure that: 1. the scroll bar is displayed immediately when the input content height exceeds the viewport height (test case in testTextArea of ​​the test module) 2. the scroll bar is highlighted when the mouse hovers over the scrollable content.The benefit of this is that it always indicates to the user that the content is not fully displayed and can be scrolled.

shun2wang commented on QMLComponents/components/JASP/Controls/DropDown.qml:
```diff
@@ -187,39 +186,71 @@ ComboBoxBase
-					border.color:		jaspTheme.focusBorderColor
-					border.width:		2
-					color:				"transparent"
+					anchors
+					{
+						top:	parent.top
+						left:	parent.left
+						right:	parent.right
+					}
+					
+					upsideDown:	true
+					extraSpace:	popupView.contentY
+				}
+				
+				ScrollMoreIndicator
+				{
+					anchors
+					{
+						left:	parent.left
+						right:	parent.right
+						bottom:	parent.bottom
+					}
+					extraSpace:	popupView.contentHeight - popupView.contentY + popupView.height 
```
A simple fix to drop down?

```suggestion
					anchors
					{
						top:	parent.top
						left:	parent.left
						right:	parent.right
					}

					upsideDown:	true
					extraSpace:	popupView.contentY
					visible:	popupView.contentY > 0
				}

				ScrollMoreIndicator
				{
					anchors
					{
						left:	parent.left
						right:	parent.right
						bottom: parent.bottom
					}
					extraSpace: popupView.contentHeight - popupView.contentY - popupView.height
					visible:	popupView.contentHeight > popupView.height
				}
```

JorisGoosen replied:
Maybe thats too simple :p

I missed some parentheses 🤡

JorisGoosen committed:
fix scroll indicator on dropdown

JorisGoosen commented:
> 1. The gradient you defined seems to be from the middle to the sides😜. I think it would be better to have a gradient from one side to the other, and the color gets darker as it gets closer to the bottom (top).

Uhm?

This is what I see:
<img width="445" alt="image" src="https://github.com/user-attachments/assets/5b08a1d8-2a0e-4d95-a1d4-fa20172cf0d3" />
(so, darker to the top/bottom depending on where it is)

Where you see it otherwise I dont know?

> 
>     2. Should we also ensure that: 1. the scroll bar is displayed immediately when the input content height exceeds the viewport height (test case in testTextArea of ​​the test module) 2. the scroll bar is highlighted when the mouse hovers over the scrollable content.The benefit of this is that it always indicates to the user that the content is not fully displayed and can be scrolled.

I like it!

Edit: I do like it, but it would require going through all the flickables JASPScrollBar is now linked to (through 'flickable'), but 'Flickable' does not tell us whether a mouse is hovering, so we would need to find a way around that.

A way would be to create a "ScrollMoreIndicator" that is not just the little gradient we made now.
But it is rather something that covers the entire "scrollable thing" and adds the gradients at top and bottom.
And it also includes a MouseArea that only checks for "containsMouse" with "hoverEnabled: true" and has "acceptedButtons: false".
This could then be a property we can link directly to the scrollbars.

While this is simple enough I dont have the time to get into that right now.
If you feel like it I can review your work!

However, I think this PR is ready to be merged now and will be already better than what we had before
